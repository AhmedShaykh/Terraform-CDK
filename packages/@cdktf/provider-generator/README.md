# @cdktf/provider-generator

`@cdktf/provider-generator` exposes an API to generate Terraform CDK provider bindings.

## Usage

```sh
yarn install @cdktf/hcl2cdk
```

### Generate bindings

```ts

```
