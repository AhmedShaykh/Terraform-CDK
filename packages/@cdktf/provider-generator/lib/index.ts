export * from "./get/constructs-maker";
export * as config from "./config";
