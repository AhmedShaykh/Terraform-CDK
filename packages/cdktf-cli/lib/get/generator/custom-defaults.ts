export const CUSTOM_DEFAULTS: { [fullName: string]: any } = {
    "aws.security_group.ingress.cidr_blocks": null,
    "aws.security_group.ingress.description": null,
    "aws.security_group.ingress.from_port": null,
    "aws.security_group.ingress.ipv6_cidr_blocks": null,
    "aws.security_group.ingress.prefix_list_ids": null,
    "aws.security_group.ingress.protocol": null,
    "aws.security_group.ingress.security_groups": null,
    "aws.security_group.ingress.self": null,
    "aws.security_group.ingress.to_port": null,
    "aws.security_group.egress.cidr_blocks": null,
    "aws.security_group.egress.description": null,
    "aws.security_group.egress.from_port": null,
    "aws.security_group.egress.ipv6_cidr_blocks": null,
    "aws.security_group.egress.prefix_list_ids": null,
    "aws.security_group.egress.protocol": null,
    "aws.security_group.egress.security_groups": null,
    "aws.security_group.egress.self": null,
    "aws.security_group.egress.to_port": null,

    "aws.default_route_table.route.cidr_block": null,
    "aws.default_route_table.route.ipv6_cidr_block": null,
    "aws.default_route_table.route.egress_only_gateway_id": null,
    "aws.default_route_table.route.gateway_id": null,
    "aws.default_route_table.route.instance_id": null,
    "aws.default_route_table.route.nat_gateway_id": null,
    "aws.default_route_table.route.network_interface_id": null,
    "aws.default_route_table.route.transit_gateway_id": null,
    "aws.default_route_table.route.vpc_endpoint_id": null,
    "aws.default_route_table.route.vpc_peering_connection_id": null,

    "aws.emr_cluster.step.action_on_failure": null,
    "aws.emr_cluster.step.hadoop_jar_step": null,
    "aws.emr_cluster.step.name": null,
    "aws.emr_cluster.step.hadoop_jar_step.args": null,
    "aws.emr_cluster.step.hadoop_jar_step.jar": null,
    "aws.emr_cluster.step.hadoop_jar_step.main_class": null,
    "aws.emr_cluster.step.hadoop_jar_step.properties": null,

    "aws.network_acl.ingress.from_port": null,
    "aws.network_acl.ingress.to_port": null,
    "aws.network_acl.ingress.rule_no": null,
    "aws.network_acl.ingress.action": null,
    "aws.network_acl.ingress.protocol": null,
    "aws.network_acl.ingress.cidr_block": null,
    "aws.network_acl.ingress.ipv6_cidr_block": null,
    "aws.network_acl.ingress.icmp_type": null,
    "aws.network_acl.ingress.icmp_code": null,
    "aws.network_acl.egress.from_port": null,
    "aws.network_acl.egress.to_port": null,
    "aws.network_acl.egress.rule_no": null,
    "aws.network_acl.egress.action": null,
    "aws.network_acl.egress.protocol": null,
    "aws.network_acl.egress.cidr_block": null,
    "aws.network_acl.egress.ipv6_cidr_block": null,
    "aws.network_acl.egress.icmp_type": null,
    "aws.network_acl.egress.icmp_code": null,

    "aws.route_table.route.cidr_block": null,
    "aws.route_table.route.ipv6_cidr_block": null,
    "aws.route_table.route.egress_only_gateway_id": null,
    "aws.route_table.route.gateway_id": null,
    "aws.route_table.route.instance_id": null,
    "aws.route_table.route.nat_gateway_id": null,
    "aws.route_table.route.local_gateway_id": null,
    "aws.route_table.route.network_interface_id": null,
    "aws.route_table.route.transit_gateway_id": null,
    "aws.route_table.route.vpc_endpoint_id": null,
    "aws.route_table.route.vpc_peering_connection_id": null,

    "aws.default_security_group.ingress.cidr_blocks": null,
    "aws.default_security_group.ingress.description": null,
    "aws.default_security_group.ingress.from_port": null,
    "aws.default_security_group.ingress.ipv6_cidr_blocks": null,
    "aws.default_security_group.ingress.prefix_list_ids": null,
    "aws.default_security_group.ingress.protocol": null,
    "aws.default_security_group.ingress.security_groups": null,
    "aws.default_security_group.ingress.self": null,
    "aws.default_security_group.ingress.to_port": null,
    "aws.default_security_group.egress.cidr_blocks": null,
    "aws.default_security_group.egress.description": null,
    "aws.default_security_group.egress.from_port": null,
    "aws.default_security_group.egress.ipv6_cidr_blocks": null,
    "aws.default_security_group.egress.prefix_list_ids": null,
    "aws.default_security_group.egress.protocol": null,
    "aws.default_security_group.egress.security_groups": null,
    "aws.default_security_group.egress.self": null,
    "aws.default_security_group.egress.to_port": null,

    "azuread.application.app_role.id": null,
    "azuread.application.app_role.allowed_member_types": null,
    "azuread.application.app_role.description": null,
    "azuread.application.app_role.display_name": null,
    "azuread.application.app_role.is_enabled": null,
    "azuread.application.app_role.value": null,
    "azuread.application.oauth2_permissions.id": null,
    "azuread.application.oauth2_permissions.admin_consent_description": null,
    "azuread.application.oauth2_permissions.admin_consent_display_name": null,
    "azuread.application.oauth2_permissions.value": null,
    "azuread.application.oauth2_permissions.type": null,
    "azuread.application.oauth2_permissions.is_enabled": null,
    "azuread.application.oauth2_permissions.user_consent_description": null,
    "azuread.application.oauth2_permissions.user_consent_display_name": null,

    "azurestack.route_table.route.name": null,
    "azurestack.route_table.route.address_prefix": null,
    "azurestack.route_table.route.next_hop_type": null,
    "azurestack.route_table.route.next_hop_in_ip_address": null,

    "vault.auth_backend.tune.default_lease_ttl": null,
    "vault.auth_backend.tune.max_lease_ttl": null,
    "vault.auth_backend.tune.audit_non_hmac_request_keys": null,
    "vault.auth_backend.tune.audit_non_hmac_response_keys": null,
    "vault.auth_backend.tune.listing_visibility": null,
    "vault.auth_backend.tune.passthrough_request_headers": null,
    "vault.auth_backend.tune.allowed_response_headers": null,
    "vault.auth_backend.tune.token_type": null,

    "vault.okta_auth_backend.group.group_name": null,
    "vault.okta_auth_backend.group.policies": null,
    "vault.okta_auth_backend.user.username": null,
    "vault.okta_auth_backend.user.groups": null,
    "vault.okta_auth_backend.user.policies": null,

    "google.compute_instance.guest_accelerator.count": null,
    "google.compute_instance.guest_accelerator.type": null,

    "google.compute_instance_from_template.attached_disk.source": null,
    "google.compute_instance_from_template.attached_disk.device_name": null,
    "google.compute_instance_from_template.attached_disk.mode": null,
    "google.compute_instance_from_template.attached_disk.disk_encryption_key_raw": null,
    "google.compute_instance_from_template.attached_disk.kms_key_self_link": null,
    "google.compute_instance_from_template.guest_accelerator.count": null,
    "google.compute_instance_from_template.guest_accelerator.type": null,
    "google.compute_instance_from_template.service_account.email": null,
    "google.compute_instance_from_template.service_account.scopes": null,
    "google.compute_instance_from_template.scratch_disk.interface": null,
    "google.compute_instance_from_template.network_interface.alias_ip_range.ip_cidr_range": null,
    "google.compute_instance_from_template.network_interface.alias_ip_range.subnetwork_range_name": null,
    "google.compute_instance_from_template.network_interface.access_config.nat_ip": null,
    "google.compute_instance_from_template.network_interface.access_config.public_ptr_domain_name": null,
    "google.compute_instance_from_template.network_interface.access_config.network_tier": null,

    "google.compute_subnetwork.secondary_ip_range.range_name": null,
    "google.compute_subnetwork.secondary_ip_range.ip_cidr_range": null,

    "google.container_cluster.guest_accelerator.count": null,
    "google.container_cluster.guest_accelerator.type": null,
    "google.container_cluster.taint.key": null,
    "google.container_cluster.taint.value": null,
    "google.container_cluster.taint.effect": null,

    "google.composer_environment.ip_allocation_policy.use_ip_aliases": null,
    "google.composer_environment.ip_allocation_policy.cluster_secondary_range_name": null,
    "google.composer_environment.ip_allocation_policy.services_secondary_range_name": null,
    "google.composer_environment.ip_allocation_policy.cluster_ipv4_cidr_block": null,
    "google.composer_environment.ip_allocation_policy.services_ipv4_cidr_block": null,
    
    "azurerm.storage_account_network_rules.bypass": null,
    "azurerm.storage_account_network_rules.ip_rules": null,
    "azurerm.storage_account_network_rules.virtual_network_subnet_ids": null,
}