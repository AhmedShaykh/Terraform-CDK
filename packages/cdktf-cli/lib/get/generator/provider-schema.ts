import * as fs from 'fs-extra';
import * as path from 'path';
import { promisify } from 'util';
import { exec, withTempDir } from '../../util';

const writeFile = promisify(fs.writeFile);
const terraformBinaryName = process.env.TERRAFORM_BINARY_NAME || 'terraform'

export interface ProviderSchema {
  format_version: '1.0';
  provider_schemas?: { [type: string]: Provider };
}

export interface Provider {
  provider: Schema;
  resource_schemas: { [type: string]: Schema };
  data_source_schemas: { [type: string]: Schema };
}

export interface Schema {
  version: number;
  block: Block;
}

export interface Attribute {
  type: AttributeType;
  description?: string;
  required?: boolean;
  optional?: boolean;
  computed?: boolean;
  sensitive?: boolean;
}

export type AttributeType =
  'string'
  | 'bool'
  | 'number'
  | [ 'set', AttributeType ]
  | [ 'map', AttributeType ]
  | [ 'list', AttributeType ]
  | [ 'object', { [attribute: string]: AttributeType } ];

export type BlockType = {
  nesting_mode: 'single' | 'map';
  block: Block;
} | {
  nesting_mode: 'list' | 'set';
  block: Block;
  min_items?: number;
  max_items?: number;
}

export interface Block {
  attributes: { [name: string]: Attribute };
  block_types: { [name: string]: BlockType };
}

export async function readSchema(providers: string[]): Promise<ProviderSchema> {
  const provider: { [name: string]: {} } = {};
  const requiredProviders: { [name: string]: { source?: string;  version?: string } } = { };

  for (const p of providers) {
    const [ fqname, version ] = p.split('@');
    const name = fqname.split('/').pop()
    if (!name) { throw new Error(`Provider name should be properly set in ${p}`) }

    provider[name] = {};
    requiredProviders[name] = { version, source: fqname };
  }
  let schema = '';

  await withTempDir('fetchSchema', async () => {
    const outdir = process.cwd();
    const filePath = path.join(outdir, 'providers.tf.json');
    // eslint-disable-next-line @typescript-eslint/camelcase
    await writeFile(filePath, JSON.stringify({ provider, terraform: { required_providers: requiredProviders }}));

    // todo: when implementing logging, we need to make sure we can show the terraform init
    // output if the log level is set to debug
    await exec(terraformBinaryName, [ 'init' ], { cwd: outdir });
    schema = await exec(terraformBinaryName, ['providers', 'schema', '-json'], { cwd: outdir });
    fs.unlinkSync(filePath)
  })

  return JSON.parse(schema);
}
