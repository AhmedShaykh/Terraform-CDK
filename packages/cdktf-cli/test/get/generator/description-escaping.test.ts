import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { TerraformProviderGenerator } from '../../../lib/get/generator/provider-generator';
import { CodeMaker } from 'codemaker';

test('broken attribute description comments', async () => {
  const code = new CodeMaker()
  const workdir = fs.mkdtempSync(path.join(os.tmpdir(), 'description-escaping.test'));
  const spec = JSON.parse(fs.readFileSync(path.join(__dirname, 'fixtures', 'description-escaping.test.fixture.json'), 'utf-8'));
  new TerraformProviderGenerator(code, spec);
  await code.save(workdir);

  const output = fs.readFileSync(path.join(workdir, 'providers/google/description-escaping.ts'), 'utf-8');
  expect(output).toMatchSnapshot();
});
