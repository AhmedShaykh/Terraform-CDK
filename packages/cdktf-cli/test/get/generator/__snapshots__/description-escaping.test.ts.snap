// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`broken attribute description comments 1`] = `
"// https://www.terraform.io/docs/providers/google/r/description_escaping.html
// generated from terraform resource schema

import { Construct } from 'constructs';
import * as cdktf from 'cdktf';

// Configuration

export interface DescriptionEscapingConfig extends cdktf.TerraformMetaArguments {
  /** The resource name of the Cloud KMS CryptoKey to be used to protect access 
 to messages published on this topic. Your project's PubSub service account 
 ('service-{{PROJECT_NUMBER}}@gcp-sa-pubsub.iam.gserviceaccount.com') must have 
 'roles/cloudkms.cryptoKeyEncrypterDecrypter' to use this feature. 
 The expected format is 'projects/*\\\\/locations/*\\\\/keyRings/*\\\\/cryptoKeys/*' */
  readonly brokenComments: boolean;
}

/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/description_escaping.html description_escaping}
*/
export class DescriptionEscaping extends cdktf.TerraformResource {

  // ===========
  // INITIALIZER
  // ===========

  /**
  * Create a new {@link https://www.terraform.io/docs/providers/google/r/description_escaping.html description_escaping}
  *
  * @param scope The scope in which to define this construct
  * @param id The scoped construct ID. Must be unique amongst siblings
  * @param options DescriptionEscapingConfig
  */
  public constructor(scope: Construct, id: string, config: DescriptionEscapingConfig) {
    super(scope, id, {
      terraformResourceType: 'description_escaping',
      terraformGeneratorMetadata: {
        providerName: 'google'
      },
      provider: config.provider,
      dependsOn: config.dependsOn,
      count: config.count,
      lifecycle: config.lifecycle
    });
    this._brokenComments = config.brokenComments;
  }

  // ==========
  // ATTRIBUTES
  // ==========

  // broken_comments - computed: false, optional: false, required: true
  private _brokenComments: boolean;
  public get brokenComments() {
    return this.getBooleanAttribute('broken_comments');
  }
  public set brokenComments(value: boolean) {
    this._brokenComments = value;
  }
  // Temporarily expose input value. Use with caution.
  public get brokenCommentsInput() {
    return this._brokenComments
  }

  // =========
  // SYNTHESIS
  // =========

  protected synthesizeAttributes(): { [name: string]: any } {
    return {
      broken_comments: cdktf.booleanToTerraform(this._brokenComments),
    };
  }
}
"
`;
